### PR TITLE
feat(components): add labels to gcp ml engine train/batch_predict jobs. Fixes #5038

### LINKED ISSUE #5038: ml_engine does not expose "labels"
I am writing kubeflow pipeline with Ai platform train job, but can't add argument "labels" (e.g. tracking costs) into ML_ENGINE_TRAIN_OP since there is no such thing. I can add labels via gcloud, REST or even Airflow has one (https://github.com/apache/airflow/blob/7f4c88c0680b4fb98fe8b31800a93e1d0476c4db/airflow/providers/google/cloud/operators/mlengine.py#L1072)

```
ML_ENGINE_TRAIN_OP = comp.load_component_from_url(
    'https://raw.githubusercontent.com/kubeflow/pipelines/1.3.0/components/gcp/ml_engine/train/component.yaml')
```

### PR DESCRIPTION
**Pull request by dkajtoch**
**Description of your changes:**
Add labels to organize jobs on ml engine 

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

**Comment by dkajtoch** [comment]
@googlebot I signed it!

File: components/gcp/ml_engine/train/component.yaml [hunk 1]
```diff
@@ -96,6 +96,12 @@ inputs:
       Defaults to 30.'
     default: '30'
     type: Integer
+  - name: labels
+    description: >-
+      One or more labels that you can add, to organize your jobs. Each label is a key-value pair,
+      where both the key and the value are arbitrary strings that you supply.
+    default: ''
+    type: Dict
```
Comment (Ark-kun) [inline components/gcp/ml_engine/train/component.yaml:104]: Let's use the `JsonObject` canonical type name here.

**Comment by Ark-kun** [comment]
How did you verify that the updated component code works?

**Comment by Ark-kun** [comment]
Interesting. I do not fully understand why this code works.
The command-line receives the labels as a JSON-serialized string. The `train` function has an untyped parameter `labels` which is treated as a dictionary. Where is the JSON stricg converted to dictionary?

**Comment by dkajtoch** [comment]
@Ark-kun I need to check it once again since I am passing `Dict` in tests and in pipeline so maybe there is no conversion. BTW in ml engine train yaml we have `List` as args so do not know why `json.dumps` is ok here https://github.com/kubeflow/pipelines/blob/master/samples/core/ai_platform/ai_platform.ipynb (I pass list directly and it works)

**Comment by Ark-kun** [comment]
> BTW in ml engine train yaml we have List as args so do not know why json.dumps is ok here

When you pass a string, the SDK assumes that you've serialized the argument yourself. Otherwise it will try to serialize the argument based on the input type. This only works for a small number of types: int, float, bool, list (==JsonArray) and dict (==JsonObject).

In the end the command line is always an array of strings when being executed.

I'd expect that the train function receives the labels as a JSON string. But then it's strange that train backend does not complain about the string instead of map.

**Comment by dkajtoch** [comment]
@Ark-kun sorry for such a delay. I did some tests with labels and it seems to work both with Dict and Json string. I added logger to `_train.py` in order to see what type I have in `train`(for labels) function. I pass labels to poc_pipeline in two forms:
```
labels = {"name": "a", "env": "dev"}  # dict
labels = json.dumps({"name": "a", "env": "dev"})  # string
```
In both cases, type is `Dict`.

I am running kubeflow pipelines via gcp ai platform version `1.0.4`

**Comment by Ark-kun** [comment]
Thank you for verifying.

I did some investigation and it looks like the Fire library is doing this. It tries to parse every argument by adding missing quotes then parsing it as python code. Since most valid JSON is also valid python ([not always](https://stackoverflow.com/questions/6627635/is-json-syntax-a-strict-subset-of-python-syntax)), this usually works.

I do not really like this - we're basically parsing JSON as if it was Python.

Could you please add explicit parsing to train and batch_predict? 

```python
import json

...

@decorators.SetParseFns(python_version=str, runtime_version=str, labels=json.loads)
def train(project_id,
```
```python

import json
from fire import decorators
...
@decorators.SetParseFns(labels=json.loads)
def batch_predict
```

https://github.com/google/python-fire/blob/c1266d0dbb2114514fcf8be62044344b5a51c733/fire/parser.py#L78

**Pull request closed without merging**